### PR TITLE
migrate Fold, Sliding, Grouped to GraphStage

### PR DESCRIPTION
**Pull request by ktoso**
This is a rebase of #20336 
Thanks a lot for the contrib @zhxiaogg !

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -338,23 +332,46 @@ private[akka] final case class Scan[In, Out](zero: Out, f: (Out, In) ⇒ Out) ex
...
+        } else {
+          pull(in)
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if (isAvailable(out)) {
+          push(out, aggregator)
+          completeStage()
+        }
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: doesn't this need a `completeStage()` in the else branch?

Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: it's completed in `onPull`
an alternative is to use `emit` ;-)

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: oh I see, thanks

**Comment by patriknw** [comment]
LGTM

**Pull request merged**